src/com/sap/piper/cm/ChangeManagement.groovy

```diff
@@ -172,29 +172,150 @@ public class ChangeManagement implements Serializable {
     void uploadFileToTransportRequestCTS(
         Map docker,
         String transportRequestId,
-        String filePath,
         String endpoint,
-        String credentialsId,
-        String cmclientOpts = '') {
+        String client,
+        String applicationName,
+        String description,
+        String abapPackage, // "package" would be better, but this is a keyword
+        String osDeployUser,
+        def deployToolDependencies,
+        def npmInstallOpts,
+        String deployConfigFile,
+        String credentialsId) {
 
-        def args = [
-                '-tID', transportRequestId,
-                "\"$filePath\""
-            ]
+        def script = this.script
 
-        int rc = executeWithCredentials(
-            BackendType.CTS,
-            docker,
-            endpoint,
-            credentialsId,
-            'upload-file-to-transport',
-            args,
-            false,
-            cmclientOpts) as int
+        def desc = description ?: 'Deployed with Piper based on SAP Fiori tools'
 
-        if(rc != 0) {
-            throw new ChangeManagementException(
-                "Cannot upload file into transport request. Return code from  cm client: $rc.")
+        if (deployToolDependencies in List) {
+            deployToolDependencies = deployToolDependencies.join(' ')
+        }
+
+        if (npmInstallOpts in List) {
+            npmInstallOpts = npmInstallOpts.join(' ')
+        }
+
+        deployToolDependencies = deployToolDependencies.trim()
+
+        /*
+            In case the configuration has been adjusted so that no deployToolDependencies are provided
+            we assume an image is used which contains already all dependencies.
+            In this case we don't invoke npm install and we run the image with the standard user
+            already, since there is no need for being root. Hence we don't need to switch user also
+            in the script.
+         */
+        boolean noInstall = deployToolDependencies.isEmpty()
+
+        Iterable cmd = ['#!/bin/bash -e']
+
+        if (! noInstall) {
+            cmd << "npm install --global ${npmInstallOpts} ${deployToolDependencies}"
+            cmd << "su ${osDeployUser}"
+        } else {
+            script.echo "[INFO] no deploy dependencies provided. Skipping npm install call. Assuning docker image '${docker?.image}' contains already the dependencies for performing the deployment."
```

Comment:
What about the case where docker is not used and the dependencies are installed on the machine; Does this echo still make sense then?